### DISCUSSION TITLE
What is a nice way to use the result of a cypress request in subsequent cypress commands?

### DISCUSSION THREAD

**Discussion by wilcoschoneveld** [cypress-io/cypress#18045] [Questions and Help] [answered]
**Introduction**
We start each test with generating test data and have created an endpoint for this. The generated test data is random but we have the option to provide parameters to the endpoint so that the data can be specified. This is nice because the test function shows exactly where the tested data comes from.

The endpoint returns the generated data and additional information, such as the id of the new data. We require this to navigate to the correct page and run the assertions. Generating the data is almost always the first thing that we do in a test.

_Example:_
```javascript
it('should show an error if highlights fail to load', () => {
    cy.request('POST', Cypress.env('API_URL') + '/generate-registration', {
        contents: {
            persons__count: 2,
            persons__1__family_name: "Johnson"
        }
    }).then((response) => {
        const { id } = response.body;

        // Go to registration
        cy.login();
        cy.visit(`/registration/${id}`);

        // Wait for highlights to be loaded
        cy.get('.persons').should('contain.text', 'Johnson');
    });
});
```

**Question**
The above is nice and works very well for us, but unfortunately it doesn't scale very well. Sometimes we need to call the generate endpoint multiple times and we don't want to be nesting the `.then()` calls if we need the results.

_Example:_
```javascript
// We want to avoid doing this...
it('should show multiple registrations', () => {
    cy.request('POST', Cypress.env('API_URL') + '/generate-registration').then((response1) => {
        cy.request('POST', Cypress.env('API_URL') + '/generate-registration').then((response2) => {
            cy.request('POST', Cypress.env('API_URL') + '/generate-registration').then((response3) => {
                // commands
                // ..
                // ..      
            })
        })
    });
});
```

It would be nice to use something like the `async`/`await` syntax provided by the javascript language to avoid nesting, like so:

```javascript
// This would be nice but it doesn't work
it('should show an error when highlights fail in a registration', async () => {
    const response = await cy.request('POST', Cypress.env('API_URL') + '/generate-registration');

    const { id } = response.body;

    cy.intercept('GET', `/registrations/${id}/highlights`, { statusCode: 500 });

    cy.login();
    cy.visit(`/registration/${id}`);

    cy.get('.warning-message').should('exist');
});

// This would be nice but it doesn't work
it('should do something cool with multiple registrations', async () => {
    const response1 = await cy.request('POST', Cypress.env('API_URL') + '/generate-registration')
    const response2 = await cy.request('POST', Cypress.env('API_URL') + '/generate-registration')
    const response3 = await cy.request('POST', Cypress.env('API_URL') + '/generate-registration')
    
    // commands
    // ..
    // ..      
});
```

Unfortunately, cypress throws warnings and errors when I try something like this, probably for good reasons. Is there a best practice that we could use here? All ideas are welcome!

Thanks a lot!

**Comment by bahmutov** [comment]
When making requests do you need to use the response from the first in the second or third one? Or are they all separate and you use the results inside the test itself?

**Reply by wilcoschoneveld** [reply]
They are all separate and I would like to use the results inside the test itself.

**Comment by bahmutov** [comment, marked as answer]
Use beforeEach hook, save as aliases and use “this.alias” inside the test itself?

**Reply by wilcoschoneveld** [reply]
Thanks for your reply! This might work. We want to have a different generate code for each test, most of the time. But I see we could do a nested `describe` with a `beforeEach` perhaps? Maybe with `context(() => {`?

**Comment by bahmutov** [comment]
Yes, you can have nested describe blocks as many levels as you want, and context is just an alias to describe so there is no difference which one to use.

**Reply by wilcoschoneveld** [reply]
Awesome! I just worked on some tests and this is workable for us, thanks again!
